test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java

```diff
@@ -700,15 +705,21 @@ private static void assertListEquals(List<Object> expected, List<Object> actual)
      * for byte[] through Arrays.equals(byte[], byte[])
      */
     @SuppressWarnings("unchecked")
-    private static void assertObjectEquals(Object expected, Object actual) {
+    private static void assertObjectEquals(Object expected, Object actual, boolean convertBase64Strings) {
         if (expected instanceof Map) {
             assertThat(actual, instanceOf(Map.class));
-            assertMapEquals((Map<String, Object>) expected, (Map<String, Object>) actual);
+            assertMapEquals((Map<String, Object>) expected, (Map<String, Object>) actual, convertBase64Strings);
         } else if (expected instanceof List) {
-            assertListEquals((List<Object>) expected, (List<Object>) actual);
+            assertListEquals((List<Object>) expected, (List<Object>) actual, convertBase64Strings);
         } else if (expected instanceof byte[]) {
             //byte[] is really a special case for binary values when comparing SMILE and CBOR, arrays of other types
             //don't need to be handled. Ordinary arrays get parsed as lists.
+            if (convertBase64Strings && actual instanceof String) {
+                // for objects deserialized from XContent the content might be a Base64 encoded string
+                actual = Base64.getDecoder().decode((String)actual);
```

Comment:
@javanna I am not entirely sure if this is the best way to do this. The other option I can think of is to always serialize to `XContent` and then assert equality.

Reply:
indeed, I am not thrilled about this part. Can you remind why we need this? What part of WriteableIngestDocument causes issues? Also confused because you didn't replace the existing assertions but you only added new ones. I wonder if this method is not testing things properly.

Reply:
So the problem is there are byte arrays in the documents (`_source` field). When you serialize it to XContent (if it is not `CBOR`) it gets serialized to a base64 string. Now when you deserialize this its hard to check for equality for these two. You are also right kind of that this method has a bug. It checks for `expected instanceof byte[]` and then casts `actual` to `byte[]` which can throw an exception. This is why I added the other assertion in the else clause. I think this is also true in the  condition for `List`.

Reply:
Before this change, I had separate tests for documents with byte arrays and without (similar to what we do with exceptions). Then you said maybe I should do `assertEqualInstances` instead of separate tests. So that led to this ;)

Reply:
ok I see, what I don't follow is why we have this problem here and not in other places, as we have other API that can also hold _source in the same way. I will need to dig this up.

Reply:
I got it. Look at GetResponseTests, it holds the source but most of the problems (what makes the test a bit special) is the fields part rather than source. The trick is to generate the source using RandomObjects.randomSource which doesn't add byte arrays, much easier for testing xcontent stuff. RandomDocumentPicks#randomSource is used for a lot of the extensive ingest unit tests but outside of xcontent serialization. I think if you change that you will not need all these changes.